src/main/java/com/derongan/minecraft/deeperworld/extensions/DeeperWorldExtensions.kt

```diff
@@ -0,0 +1,60 @@
+package com.derongan.minecraft.deeperworld.extensions
+
+import org.bukkit.Material
+import org.bukkit.entity.Entity
+import org.bukkit.entity.Item
+import org.bukkit.entity.LivingEntity
+import org.bukkit.entity.Player
+
+internal fun Entity.getVehicleRecursive(): Entity? {
+    var currentVehicle = vehicle
+
+    while (currentVehicle?.vehicle != null) {
+        currentVehicle = currentVehicle.vehicle
+    }
+
+    return currentVehicle
+}
+
+internal fun Entity.getPlayerPassengersRecursive(): List<Player>? {
+    return mutableListOf<Player>().let { playerList ->
+        if (passengers.size != 0) {
+            if (passengers.any { it is Player }) {
+                playerList.addAll(passengers.filterIsInstance<Player>())
+            }
+
+            for (passenger: Entity in passengers) {
+                passenger.getPlayerPassengersRecursive()?.let { playerList.addAll(it) }
+            }
+        }
+
+        if (playerList.isNotEmpty()) {
+            return@let playerList
+        } else {
+            return@let null
+        }
+    }
+}
+
+internal fun Player.getLeashedEntities(): List<LivingEntity>? {
+    return getNearbyEntities(20.0, 20.0, 20.0)
+            .filter { it is LivingEntity && (it.isLeashed && it.leashHolder == this) }
```

Comment:
You can use:
```kotlin
            .filterIsInstance<LivingEntity>()
            .filter { it.isLeashed && it.leashHolder == this }
```
Then don't need to use the cast afterwards. 

I also think it might be simpler to return an empty list instead of null when there are no entities leashed. Same goes for the other functions.

Reply:
Yeah fair, I had that before but thought optimization wise 1 filter is better than 2. But with the cast in the mix I guess it doesn't matter anyway. I'll change it.

About the returning null instead of an empty list, it would be simpler but in my mind it makes more sense to return null if there are no results. But this might be bias from other language constructs. I see that standard functions like "getNearbyEntities" also always returns a list, even if there are no entities. So I'll keep that convention as well then.